ports/espressif/common-hal/wifi/Monitor.c

```diff
@@ -0,0 +1,166 @@
+
+    // prepare packet
+    monitor_packet_t packet = {
+        .channel = pkt->rx_ctrl.channel,
+        .length = pkt->rx_ctrl.sig_len,
+        .rssi = pkt->rx_ctrl.rssi,
+    };
+
+    // for now, the monitor only dumps the length of the MISC type frame
+    if (type != WIFI_PKT_MISC && !pkt->rx_ctrl.rx_state) {
+        packet.length -= MONITOR_PAYLOAD_FCS_LEN;
+        packet.payload = malloc(packet.length);
+        if (packet.payload) {
+            memcpy(packet.payload, pkt->payload, packet.length);
+            wifi_monitor_obj_t *self = MP_STATE_VM(wifi_monitor_singleton);
+            if (self->queue) {
+                // send packet
+                if (xQueueSendFromISR(self->queue, &packet, NULL) != pdTRUE) {
+                    self->loss++;
+                    free(packet.payload);
+                    ESP_LOGE(TAG, "packet queue full");
+                }
+            }
+        } else {
+            ESP_LOGE(TAG, "not enough memory for packet");
+        }
+    }
+}
+
+void common_hal_wifi_monitor_construct(wifi_monitor_obj_t *self, uint8_t channel, size_t queue) {
+    const compressed_string_t *monitor_mode_init_error = translate("monitor init failed");
+
+    self->queue = xQueueCreate(queue, sizeof(monitor_packet_t));
+    if (!self->queue) {
+        mp_raise_RuntimeError(monitor_mode_init_error);
+    }
+
+    // start wifi promicuous mode
+    wifi_promiscuous_filter_t wifi_filter = {
+        .filter_mask = WIFI_PROMIS_FILTER_MASK_MGMT,
+    };
+    esp_wifi_set_promiscuous_filter(&wifi_filter);
+    esp_wifi_set_promiscuous_rx_cb(wifi_monitor_cb);
+    if (esp_wifi_set_promiscuous(true) != ESP_OK) {
+        mp_raise_RuntimeError(monitor_mode_init_error);
+    }
+    esp_wifi_set_channel(channel, WIFI_SECOND_CHAN_NONE);
+
+    self->channel = channel;
+    self->queue_length = queue;
+}
+
+bool common_hal_wifi_monitor_deinited(void) {
+    bool enabled;
+    esp_wifi_get_promiscuous(&enabled);
+    return !enabled;
+}
+
+void common_hal_wifi_monitor_deinit(wifi_monitor_obj_t *self) {
+    if (common_hal_wifi_monitor_deinited()) {
+        return;
+    }
+
+    // disable wifi promiscuous mode
+    esp_wifi_set_promiscuous(false);
+
+    // make sure to free all resources in the left items
+    UBaseType_t left_items = uxQueueMessagesWaiting(self->queue);
+    monitor_packet_t packet;
+    while (left_items--) {
+        xQueueReceive(self->queue, &packet, MONITOR_QUEUE_TIMEOUT_TICK);
+        free(packet.payload);
+    }
+    vQueueDelete(self->queue);
+    self->queue = NULL;
+}
+
+void common_hal_wifi_monitor_set_channel(wifi_monitor_obj_t *self, uint8_t channel) {
+    self->channel = channel;
+    esp_wifi_set_channel(channel, WIFI_SECOND_CHAN_NONE);
+}
+
+mp_obj_t common_hal_wifi_monitor_get_channel(wifi_monitor_obj_t *self) {
+    return MP_OBJ_NEW_SMALL_INT(self->channel);
+}
+
+mp_obj_t common_hal_wifi_monitor_get_queue(wifi_monitor_obj_t *self) {
+    return mp_obj_new_int_from_uint(self->queue_length);
```

Comment:
Looks like this got reverted from
`return mp_obj_new_int_from_uint(uxQueueMessagesWaiting(self->queue));`

Reply:
This is for returning `queue` property set in the constructor.
Instead of returning `uxQueueMessagesWaiting`, I chose to implement the `loss` method.

Reply:
Queue size is more informative than loss count. Queue filling is a warning sign that loss could occur due to memory or queue full. Do we need to read back the value we set if it never changes? If so, I can PR a new method for queue size once this merges (I can't seem to do a patch on this PR).

Reply:
Added a `monitor.queued()` method for this.